File: backend/app/schemas/maintenance.py
from datetime import datetime

from pydantic import BaseModel

from backend.app.domain.maintenance import BackupInfo, DiagnosticsReport


class BackupResponse(BaseModel):
    file_name: str
    path: str
    created_at: datetime
    size_bytes: int

    @classmethod
    def from_domain(cls, backup: BackupInfo) -> "BackupResponse":
        return cls(
            file_name=backup.path.name,
            path=str(backup.path),
            created_at=backup.created_at,
            size_bytes=backup.size_bytes,
        )


class DiagnosticsResponse(BaseModel):
    app_version: str
    python_version: str
    platform: str
    api_host: str
    database_path: str
    backup_directory: str
    database_size_bytes: int
    database_integrity: str
    database_revision: str | None
    workspace_count: int
    backup_count: int

    @classmethod
    def from_domain(cls, report: DiagnosticsReport) -> "DiagnosticsResponse":
        return cls(
            app_version=report.app_version,
            python_version=report.python_version,
            platform=report.platform,
            api_host=report.api_host,
            database_path=str(report.database.database_path),
            backup_directory=str(report.database.backup_directory),
            database_size_bytes=report.database.database_size_bytes,
            database_integrity=report.database.database_integrity,
            database_revision=report.database.database_revision,
            workspace_count=report.database.workspace_count,
            backup_count=report.database.backup_count,
        )

File: frontend/src/api/maintenance.ts
import { resolveBackendConnection } from "./backend-connection";
import { ApiClient, ApiClientError } from "./client";

export interface BackupInfo {
  file_name: string;
  path: string;
  created_at: string;
  size_bytes: number;
}

export interface DiagnosticsReport {
  app_version: string;
  python_version: string;
  platform: string;
  api_host: string;
  database_path: string;
  backup_directory: string;
  database_size_bytes: number;
  database_integrity: string;
  database_revision: string | null;
  workspace_count: number;
  backup_count: number;
}

function isNonNegativeNumber(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value) && value >= 0;
}

function isBackupInfo(value: unknown): value is BackupInfo {
  if (typeof value !== "object" || value === null) {
    return false;
  }
  const candidate = value as Partial<BackupInfo>;
  return (
    typeof candidate.file_name === "string" &&
    typeof candidate.path === "string" &&
    typeof candidate.created_at === "string" &&
    isNonNegativeNumber(candidate.size_bytes)
  );
}

function validateBackup(value: unknown): BackupInfo {
  if (!isBackupInfo(value)) {
    throw new ApiClientError("后端备份响应格式不正确。", { code: "INVALID_RESPONSE" });
  }
  return value;
}

function validateBackups(value: unknown): BackupInfo[] {
  if (!Array.isArray(value) || !value.every(isBackupInfo)) {
    throw new ApiClientError("后端备份列表响应格式不正确。", {
      code: "INVALID_RESPONSE",
    });
  }
  return value;
}

function validateDiagnostics(value: unknown): DiagnosticsReport {
  if (typeof value !== "object" || value === null) {
    throw new ApiClientError("后端诊断响应格式不正确。", { code: "INVALID_RESPONSE" });
  }
  const candidate = value as Partial<DiagnosticsReport>;
  const valid =
    typeof candidate.app_version === "string" &&
    typeof candidate.python_version === "string" &&
    typeof candidate.platform === "string" &&
    candidate.api_host === "127.0.0.1" &&
    typeof candidate.database_path === "string" &&
    typeof candidate.backup_directory === "string" &&
    isNonNegativeNumber(candidate.database_size_bytes) &&
    typeof candidate.database_integrity === "string" &&
    (candidate.database_revision === null || typeof candidate.database_revision === "string") &&
    isNonNegativeNumber(candidate.workspace_count) &&
    isNonNegativeNumber(candidate.backup_count);
  if (!valid) {
    throw new ApiClientError("后端诊断响应格式不正确。", { code: "INVALID_RESPONSE" });
  }
  return candidate as DiagnosticsReport;
}

async function maintenanceClient(): Promise<ApiClient> {
  const connection = await resolveBackendConnection();
  return new ApiClient(connection.baseUrl, connection.token);
}

export async function getDiagnostics(): Promise<DiagnosticsReport> {
  const client = await maintenanceClient();
  return validateDiagnostics(await client.get<unknown>("/api/diagnostics"));
}

export async function listBackups(): Promise<BackupInfo[]> {
  const client = await maintenanceClient();
  return validateBackups(await client.get<unknown>("/api/backups"));
}

export async function createBackup(): Promise<BackupInfo> {
  const client = await maintenanceClient();
  return validateBackup(await client.post<unknown>("/api/backups"));
}
